Custom projectile trajectories

Starkku opened the pull request:
This is rebased & reopened version of [secsome's original PR](https://github.com/Phobos-developers/Phobos/pull/432).

- Projectiles can now have customizable trajectories.
  - `Trajectory` should not be combined with original game's projectile trajectory logics (`Arcing`, `ROT` or `Inviso`).

#### Straight trajectory

- Self-explanatory, is a straight-shot trajectory.
  - Initial speed is determined by weapon's `Trajectory.Speed`.

In `rulesmd.ini`
```ini
[SOMEPROJECTILE]     ; Projectile
Trajectory=Straight  ; Trajectory type
```

#### Bombard trajectory

- Similar trajectory to `Straight`, but targets a coordinate above the intended target (height determined by `Trajectory.Bombard.Height`). When the projectile approaches that coordinate, it will free fall and explodes when it hits the target or ground.
  - Initial speed is determined by weapon's `Trajectory.Speed`.

In `rulesmd.ini`
```ini
[SOMEPROJECTILE]               ; Projectile
Trajectory=Bombard             ; Trajectory type
Trajectory.Bombard.Height=0.0  ; double
```

- This sets projectile speed used by custom projectile trajectories.

In `rulesmd.ini`:
```ini
[SOMEWEAPON]            ; WeaponType
Trajectory.Speed=100.0  ; double
```

AlliedG commented:
Tested and did not encounter any obvious issues 

Trajectory.Speed above 200 can lead to overshooting (easily noticeable with something like Trajectory.Speed=1000.0 but likely dependent on a variety of factors, would advise modders to keep to sensible numbers for all values.

![500-speed-100-and-5000-speed-200](https://user-images.githubusercontent.com/84816840/162592044-6d654f7b-850a-4441-a215-ca70b2b8dfda.gif)

Trajectory.Bombard.Height=500 with Trajectory.Speed=100.0 and Trajectory.Bombard.Height=5000 with Trajectory.Speed=200.0

![speed-10](https://user-images.githubusercontent.com/84816840/162592048-abddfb6c-546f-46b0-884d-6caa173771f4.gif)

Trajectory=Straight Trajectory.Speed=10

![straight](https://user-images.githubusercontent.com/84816840/162592050-9188569c-d7be-4388-b910-f6ef9f1d461a.gif)

Trajectory=Straight SubjectToCliffs=no Trajectory.Speed omitted

![straight-not-subject-to-cliffs](https://user-images.githubusercontent.com/84816840/162592052-e0fba7b0-6ff3-4b72-852e-7a6792c793c6.gif)

Trajectory=Straight Trajectory.Speed omitted

Speederovsky commented:
Doesn't explode on collision anymore. Cluster works too.

Starkku commented:
Added new functions `OnAITargetCoordCheck` and `OnAITechnoCheck` in trajectory classes that are executed before game does any checks for if bullet has reached target cell or if bullet encounters a TechnoClass instance in a cell it is on, their return values determine what is done in regards to the game checks. They can potentially be used to create more complex projectile behaviour patterns in future, but for now they simply ensure that the current trajectories do not exhibit any weird behaviour (aforementioned collision issues with enemy units which are now fixed).

Without further ago, merging as this has been tested fairly extensively by now.
